### ISSUE TITLE
Sql data source doesn't work when columns contain Datetime type.

### ISSUE THREAD

**Issue by himalpatel81** [dotnet/machinelearning-modelbuilder#1897]
**System Information (please complete the following information):**
 - Model Builder Version (available in Manage Extensions dialog):  16.7.3.2143002
 - Visual Studio Version : VS 2022

**Describe the bug**
- I'm training sql server data for value prediction. While starting training getting the error.  
- Clear description of the problem:
- 

**To Reproduce**
1. Go to Model Builder
2. Click on Value Prediction
3. Select SQL Server Data Source and any table. pick columns to project. 
4. Click Train button. 
Video: [Reproduction video](https://himal.tinytake.com/df/1162506/thumbnail?type=attachments&version_no=0&file_version_no=0&thumbnail_size=preview)

**Expected behavior**
It should successfully train the data. 

**Screenshots**

**Additional context**
2021-11-02 23:55:48.6881 INFO Set log file path to C:\Users\hpatel\AppData\Local\Temp\MLVSTools\logs\MLWorkflowPredictionModel-RWOK9E.txt (Microsoft.ML.ModelBuilder.Utils.Logger.Info)
2021-11-02 23:56:09.2491 DEBUG AutoMLType: Octopus (Microsoft.ML.ModelBuilder.Utils.Logger.Debug)
2021-11-02 23:56:09.4371 TRACE [Source=DatabaseLoader; Binding, Kind=Trace] Channel started (Microsoft.ML.ModelBuilder.ViewModels.TrainViewModel.AutoMLService_DiagnosticDataReceived)
2021-11-02 23:56:09.4541 TRACE [Source=DatabaseLoader; Binding, Kind=Trace] Channel finished. Elapsed 00:00:00.0200831. (Microsoft.ML.ModelBuilder.ViewModels.TrainViewModel.AutoMLService_DiagnosticDataReceived)
2021-11-02 23:56:09.4541 TRACE [Source=DatabaseLoader; Binding, Kind=Trace] Channel disposed (Microsoft.ML.ModelBuilder.ViewModels.TrainViewModel.AutoMLService_DiagnosticDataReceived)
2021-11-02 23:56:09.4961 INFO start nni training (Microsoft.ML.ModelBuilder.Utils.Logger.Info)
2021-11-02 23:56:09.5341 INFO Experiment output folder: C:\Users\hpatel\AppData\Local\Temp\AutoML-NNI\Experiment-VH6LNV (Microsoft.ML.ModelBuilder.Utils.Logger.Info)
2021-11-02 23:56:09.5341 DEBUG row count is unknown, count it explicitly (Microsoft.ML.ModelBuilder.Utils.Logger.Debug)
2021-11-02 23:56:09.6171 TRACE [Source=DatabaseLoader; Cursor, Kind=Trace] Channel started (Microsoft.ML.ModelBuilder.ViewModels.TrainViewModel.AutoMLService_DiagnosticDataReceived)
2021-11-02 23:56:10.4886 TRACE [Source=DatabaseLoader; Cursor, Kind=Trace] Channel finished. Elapsed 00:00:00.8720578. (Microsoft.ML.ModelBuilder.ViewModels.TrainViewModel.AutoMLService_DiagnosticDataReceived)
2021-11-02 23:56:10.4886 TRACE [Source=DatabaseLoader; Cursor, Kind=Trace] Channel disposed (Microsoft.ML.ModelBuilder.ViewModels.TrainViewModel.AutoMLService_DiagnosticDataReceived)
2021-11-02 23:56:10.4886 DEBUG count elapse 951ms (Microsoft.ML.ModelBuilder.Utils.Logger.Debug)
2021-11-02 23:56:10.5106 INFO |     Trainer                             RSquared Absolute-loss Squared-loss RMS-loss  Duration #Iteration      | (Microsoft.ML.ModelBuilder.Utils.Logger.Info)
2021-11-02 23:56:10.9862 DEBUG select pipeline e0 * e1 * e3 (Microsoft.ML.ModelBuilder.Utils.Logger.Debug)
2021-11-02 23:56:11.0362 DEBUG 2$NumberOfTrees:4,2$FeatureFraction:1 (Microsoft.ML.ModelBuilder.Utils.Logger.Debug)
2021-11-02 23:56:11.0632 DEBUG update train dataset C:\Users\hpatel\AppData\Local\Temp\AutoML-NNI\Experiment-VH6LNV\train.bin (Microsoft.ML.ModelBuilder.Utils.Logger.Debug)
2021-11-02 23:56:11.1342 INFO Trial 0 encounter error with message: Read only 0 bytes in file, expected header size of 256 (Microsoft.ML.ModelBuilder.Utils.Logger.Info)
2021-11-02 23:56:11.2672 DEBUG    at Microsoft.ML.Data.IO.BinaryLoader.InitHeader()
   at Microsoft.ML.Data.IO.BinaryLoader..ctor(Arguments args, IHost host, Stream stream, Boolean leaveOpen)
   at Microsoft.ML.BinaryLoaderSaverCatalog.LoadFromBinary(DataOperationsCatalog catalog, String path)
   at Microsoft.ML.ModelBuilder.AutoMLService.LocalDispatcher.Run() in /_/src/Microsoft.ML.ModelBuilder.AutoMLService/Dispatcher/LocalDispatcher.cs:line 115
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.ML.ModelBuilder.AutoMLService.LocalDispatcher.<RunAsync>d__19.MoveNext() in /_/src/Microsoft.ML.ModelBuilder.AutoMLService/Dispatcher/LocalDispatcher.cs:line 153
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.ML.ModelBuilder.AutoMLService.LocalDispatcher.<AddTrialAsync>d__16.MoveNext() in /_/src/Microsoft.ML.ModelBuilder.AutoMLService/Dispatcher/LocalDispatcher.cs:line 51 (Microsoft.ML.ModelBuilder.Utils.Logger.Debug)
2021-11-02 23:56:11.3332 DEBUG Object reference not set to an instance of an object.
   at Microsoft.ML.ModelBuilder.AutoMLService.Proposer.PipelineProposer.Update(TrialParameter parameter, TrialResult result) in /_/src/Microsoft.ML.ModelBuilder.AutoMLService/Proposer/PipelineProposer.cs:line 254
   at Microsoft.ML.ModelBuilder.AutoMLService.Experiments.LocalAutoMLExperiment.<ExecuteAsync>d__15.MoveNext() in /_/src/Microsoft.ML.ModelBuilder.AutoMLService/Experiments/LocalAutoMLExperiment.cs:line 161
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.ML.ModelBuilder.AutoMLEngine.<StartTrainingAsync>d__21.MoveNext() in /_/src/Microsoft.ML.ModelBuilder.AutoMLService/AutoMLEngineService/AutoMLEngine.cs:line 167   at StreamJsonRpc.JsonRpc.<InvokeCoreAsync>d__139`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.ML.ModelBuilder.ViewModels.TrainViewModel.<<StartTrainingAsync>b__109_0>d.MoveNext() (Microsoft.ML.ModelBuilder.Utils.Logger.Debug)
2021-11-02 23:58:13.1781 DEBUG AutoMLType: Octopus (Microsoft.ML.ModelBuilder.Utils.Logger.Debug)
2021-11-02 23:58:13.1841 TRACE [Source=DatabaseLoader; Binding, Kind=Trace] Channel started (Microsoft.ML.ModelBuilder.ViewModels.TrainViewModel.AutoMLService_DiagnosticDataReceived)
2021-11-02 23:58:13.1841 TRACE [Source=DatabaseLoader; Binding, Kind=Trace] Channel finished. Elapsed 00:00:00.0001717. (Microsoft.ML.ModelBuilder.ViewModels.TrainViewModel.AutoMLService_DiagnosticDataReceived)
2021-11-02 23:58:13.1841 TRACE [Source=DatabaseLoader; Binding, Kind=Trace] Channel disposed (Microsoft.ML.ModelBuilder.ViewModels.TrainViewModel.AutoMLService_DiagnosticDataReceived)
2021-11-02 23:58:13.1841 INFO start nni training (Microsoft.ML.ModelBuilder.Utils.Logger.Info)
2021-11-02 23:58:13.1841 INFO Experiment output folder: C:\Users\hpatel\AppData\Local\Temp\AutoML-NNI\Experiment-VH6LNV (Microsoft.ML.ModelBuilder.Utils.Logger.Info)
2021-11-02 23:58:13.1841 DEBUG row count is unknown, count it explicitly (Microsoft.ML.ModelBuilder.Utils.Logger.Debug)
2021-11-02 23:58:13.1841 TRACE [Source=DatabaseLoader; Cursor, Kind=Trace] Channel started (Microsoft.ML.ModelBuilder.ViewModels.TrainViewModel.AutoMLService_DiagnosticDataReceived)
2021-11-02 23:58:13.2171 TRACE [Source=DatabaseLoader; Cursor, Kind=Trace] Channel finished. Elapsed 00:00:00.0318172. (Microsoft.ML.ModelBuilder.ViewModels.TrainViewModel.AutoMLService_DiagnosticDataReceived)
2021-11-02 23:58:13.2171 TRACE [Source=DatabaseLoader; Cursor, Kind=Trace] Channel disposed (Microsoft.ML.ModelBuilder.ViewModels.TrainViewModel.AutoMLService_DiagnosticDataReceived)
2021-11-02 23:58:13.2171 DEBUG count elapse 32ms (Microsoft.ML.ModelBuilder.Utils.Logger.Debug)
2021-11-02 23:58:13.2171 INFO |     Trainer                             RSquared Absolute-loss Squared-loss RMS-loss  Duration #Iteration      | (Microsoft.ML.ModelBuilder.Utils.Logger.Info)
2021-11-02 23:58:13.2171 DEBUG select pipeline e0 * e1 * e3 (Microsoft.ML.ModelBuilder.Utils.Logger.Debug)
2021-11-02 23:58:13.2171 DEBUG 2$NumberOfTrees:4,2$FeatureFraction:1 (Microsoft.ML.ModelBuilder.Utils.Logger.Debug)
2021-11-02 23:58:13.2171 DEBUG update train dataset C:\Users\hpatel\AppData\Local\Temp\AutoML-NNI\Experiment-VH6LNV\train.bin (Microsoft.ML.ModelBuilder.Utils.Logger.Debug)
2021-11-02 23:58:13.2171 INFO Trial 0 encounter error with message: Read only 0 bytes in file, expected header size of 256 (Microsoft.ML.ModelBuilder.Utils.Logger.Info)
2021-11-02 23:58:13.2171 DEBUG    at Microsoft.ML.Data.IO.BinaryLoader.InitHeader()
   at Microsoft.ML.Data.IO.BinaryLoader..ctor(Arguments args, IHost host, Stream stream, Boolean leaveOpen)
   at Microsoft.ML.BinaryLoaderSaverCatalog.LoadFromBinary(DataOperationsCatalog catalog, String path)
   at Microsoft.ML.ModelBuilder.AutoMLService.LocalDispatcher.Run() in /_/src/Microsoft.ML.ModelBuilder.AutoMLService/Dispatcher/LocalDispatcher.cs:line 115
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.ML.ModelBuilder.AutoMLService.LocalDispatcher.<RunAsync>d__19.MoveNext() in /_/src/Microsoft.ML.ModelBuilder.AutoMLService/Dispatcher/LocalDispatcher.cs:line 153
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.ML.ModelBuilder.AutoMLService.LocalDispatcher.<AddTrialAsync>d__16.MoveNext() in /_/src/Microsoft.ML.ModelBuilder.AutoMLService/Dispatcher/LocalDispatcher.cs:line 51 (Microsoft.ML.ModelBuilder.Utils.Logger.Debug)
2021-11-02 23:58:13.2301 DEBUG Object reference not set to an instance of an object.
   at Microsoft.ML.ModelBuilder.AutoMLService.Proposer.PipelineProposer.Update(TrialParameter parameter, TrialResult result) in /_/src/Microsoft.ML.ModelBuilder.AutoMLService/Proposer/PipelineProposer.cs:line 254
   at Microsoft.ML.ModelBuilder.AutoMLService.Experiments.LocalAutoMLExperiment.<ExecuteAsync>d__15.MoveNext() in /_/src/Microsoft.ML.ModelBuilder.AutoMLService/Experiments/LocalAutoMLExperiment.cs:line 161
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.ML.ModelBuilder.AutoMLEngine.<StartTrainingAsync>d__21.MoveNext() in /_/src/Microsoft.ML.ModelBuilder.AutoMLService/AutoMLEngineService/AutoMLEngine.cs:line 167   at StreamJsonRpc.JsonRpc.<InvokeCoreAsync>d__139`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.ML.ModelBuilder.ViewModels.TrainViewModel.<<StartTrainingAsync>b__109_0>d.MoveNext() (Microsoft.ML.ModelBuilder.Utils.Logger.Debug)

**Comment by LittleLittleCloud** [comment]
Looks like trial fail to load data from dataset

![image](https://user-images.githubusercontent.com/16876986/139910446-d6b44d93-5ec2-4ccf-a505-744fcfe0405a.png)

Can you upload `C:\Users\hpatel\AppData\Local\Temp\AutoML-NNI\Experiment-VH6LNV\train.bin` so we can examine the file header, regards

**Comment by himalpatel81** [comment]
Hi @LittleLittleCloud ,
train.bin file is empty. 0 KB. 
 I restart the VS and tried again and got attached error. 

Error Detail: Shuffle input cursor reader failed with an exception. 
If I try again, the error is : Object reference not set to an instance of an object. 

[MLWorkflowPredictionModel-W41J9H.txt](https://github.com/dotnet/machinelearning-modelbuilder/files/7464684/MLWorkflowPredictionModel-W41J9H.txt)
![ErrorTraining](https://user-images.githubusercontent.com/14814661/139969292-ec905ac0-ff1e-4f20-9fa9-ef673a28e195.jpg)

Please let me know if any additional detail is required. 

Regards,
Himal

**Comment by LittleLittleCloud** [comment]
I see, so it seems that the error is because model builder fails to save IDataView into train.bin when data source is sql. We will take a look.

In the meanwhile, one work-around can be export sql as csv and load it as file. Let us know if that works for you

**Comment by LittleLittleCloud** [comment]
Hmmm I have some trouble reproducing this error, would you mind sharing the schema and maybe the first 100 row of data with us so I can reproduce it?

**Comment by himalpatel81** [comment]
Got a different error while using CSV. Attached the screenshot. Coulple of the column in my schema is datetime. not sure if that is the root cause of the error. 
Also, attached the SQL schema and the data, CSV Data. 

Please let me know if any additional detail is required
![advanced configuration](https://user-images.githubusercontent.com/14814661/140232753-76c6877b-e1d6-4cd7-a853-98a9e46a4f36.jpg)
![CSV Error](https://user-images.githubusercontent.com/14814661/140232755-2d887336-d9fe-4216-9433-6b9daae08b7a.jpg)
[CSV.csv](https://github.com/dotnet/machinelearning-modelbuilder/files/7471286/CSV.csv)
[script.zip](https://github.com/dotnet/machinelearning-modelbuilder/files/7471287/script.zip)
.

**Comment by LittleLittleCloud** [comment]
@himalpatel81 
Could be the Datetime problem. I'll see if I can reproduce it on my side.

BTW looks like the csv training is success on my side, looks like it's some wield issue in data source, where the column type is not correctly updated. Could you create a new .mbconfig and start from scratch to see if it works?
![image](https://user-images.githubusercontent.com/16876986/140556554-579223d4-f276-4e7e-a9c9-af3ab79e6eac.png)
![image](https://user-images.githubusercontent.com/16876986/140556590-a3cda42e-9f2b-40c6-b095-eaeaaed58284.png)

**Comment by LittleLittleCloud** [comment]
After some digging, looks like the error is because the column inference result from model builder doesn't actually match table schema. In the table, the type of CreatedDate and CompleteDate is Datetime, while in column inference result, both are string, which results to an error while downloading dataset.

@beccamc 
Can you take a look at this issue. I feel like the root cause is we didn't create the correct schema when inferencing column so maybe we need to add Datetime in column type to fix this? I also upload this table to sql-dataset.database.windows.net/regression/predection so you should be able to reproduce this error easily.

**Comment by beccamc** [comment]
Duplicate of https://github.com/dotnet/machinelearning-modelbuilder/issues/1744, which is on the backlog. We should absolutely fix! 

In this case though, adding date time won't unblock the customer because we can't do a value prediction on a date. Right?

**Comment by LittleLittleCloud** [comment]
> In this case though, adding date time won't unblock the customer because we can't do a value prediction on a date. Right?

Depends on how "block" is defined. If block includes user error (having datetime as label type, etc), then yes.

**Comment by blahblehblah** [comment]
I am having similar issues

Console
`Trial 5974 encounter error with message: Read only 0 bytes in file, expected header size of 256
Trial 5975 encounter error with message: Read only 0 bytes in file, expected header size of 256
Trial 5976 encounter error with message: Read only 0 bytes in file, expected header size of 256`


Logs
`2021-11-14 22:21:11.8093 DEBUG    at Microsoft.ML.Data.IO.BinaryLoader.InitHeader()
   at Microsoft.ML.Data.IO.BinaryLoader..ctor(Arguments args, IHost host, Stream stream, Boolean leaveOpen)
   at Microsoft.ML.BinaryLoaderSaverCatalog.LoadFromBinary(DataOperationsCatalog catalog, String path)
   at Microsoft.ML.ModelBuilder.AutoMLService.LocalDispatcher.Run() in /_/src/Microsoft.ML.ModelBuilder.AutoMLService/Dispatcher/LocalDispatcher.cs:line 113
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.ML.ModelBuilder.AutoMLService.LocalDispatcher.<RunAsync>d__19.MoveNext() in /_/src/Microsoft.ML.ModelBuilder.AutoMLService/Dispatcher/LocalDispatcher.cs:line 151
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.ML.ModelBuilder.AutoMLService.LocalDispatcher.<AddTrialAsync>d__16.MoveNext() in /_/src/Microsoft.ML.ModelBuilder.AutoMLService/Dispatcher/LocalDispatcher.cs:line 49 (Microsoft.ML.ModelBuilder.Utils.Logger.Debug)
`

**Comment by beccamc** [comment]
Tracking adding the DateTime type in issue #1744. Please watch for updates there.
